polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -505,6 +508,62 @@ bool usesValue(Value user, Value use) {
                       [&](Value operand) { return usesValue(operand, use); });
 }
 
+/// Return the local id ordering for \p accSub to index the shared local memory
+/// buffer created for \p accSub by this optimization. The algorithm is to use
+/// the local id associated with the global id used by a dimension. For indexes
+/// that do not reference global id, use unused local id in increasing order.
+/// For example, assuming tx and ty are the global id of dimension 0 and
+/// dimension 1,
+///   A[ty][i], global id 1 is used as the first index,
+///   so local id 1 is used as the first index, and remaining local id 0 is used
+///   as the second index. The result is [1, 0].
+///   A[tx][i], global id 0 is used as the first index,
+///   so local id 0 is used as the first index, and remaining local id 1 is used
+///   as the second index. The result is [0, 1].
+///   A[i][tx], global id 0 is used as the second index,
+///   so local id 0 is used as the second index, and remaining local id 1 is
+///   used as the first index. The result is [1, 0]. A[i][j], global ids are not
+///   used, so local id are used in increasing order. The result is [0, 1].
+SmallVector<unsigned> getLocalIDOrdering(sycl::SYCLAccessorSubscriptOp accSub,
+                                         DataFlowSolver &solver) {
+  const SmallVector<Value> indexes = getIndexes(accSub, solver);
+  auto func = accSub->getParentOfType<FunctionOpInterface>();
+  const SmallVector<Value> globalIDs = polygeist::getThreadVector(func, solver);
+  assert(globalIDs.size() == indexes.size() &&
+         "Expecting number of global id the same as number of indexes");
+
+  SmallVector<std::optional<unsigned>> localIDOrdering(globalIDs.size(),
+                                                       std::nullopt);
+  llvm::SmallSet<int, 3> unusedIndexes;
+  for (unsigned globalDim = 0; globalDim < globalIDs.size(); ++globalDim) {
+    std::list<unsigned> dimensions(indexes.size());
+    std::iota(dimensions.begin(), dimensions.end(), 0);
+    // Insert the dimension of unused local id in unusedIndexes.
+    if (llvm::none_of(dimensions, [&](unsigned dim) {
+          if (globalIDs[globalDim] &&
+              usesValue(indexes[dim], globalIDs[globalDim])) {
+            localIDOrdering[dim] = globalDim;
+            return true;
+          }
+          return false;
+        }))
+      unusedIndexes.insert(globalDim);
+  }
```

Comment:
Can we have a `for` loop instead of creating a `list` here?
Also, do we mind if more than one dimension fulfills the `usesValue` condition?

Reply:
> Can we have a for loop instead of creating a list here?

It is originally a `for` loop, it is changed due to a review comment: https://github.com/intel/llvm/pull/10053#discussion_r1240100258.

> Also, do we mind if more than one dimension fulfills the usesValue condition?

That shouldn't happen, as such access is not considered as isCandidateAccess after https://github.com/intel/llvm/pull/10056.

Reply:
:+1: